lib/evaluate/shape.h

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// GetShape() analyzes an expression and determines its shape, if possible,
+// representing the result as a vector of scalar integer expressions.
+
+#ifndef FORTRAN_EVALUATE_SHAPE_H_
+#define FORTRAN_EVALUATE_SHAPE_H_
+
+#include "expression.h"
+#include "tools.h"
+#include "type.h"
+#include "../common/indirection.h"
+#include <optional>
+#include <variant>
+
+namespace Fortran::evaluate {
+
+using ExtentType = SubscriptInteger;
+using ExtentExpr = Expr<ExtentType>;
+using MaybeExtent = std::optional<ExtentExpr>;
+using Shape = std::vector<MaybeExtent>;
+
+// Convert between various representations of shapes
+Shape AsShape(const Constant<ExtentType> &arrayConstant);
+std::optional<Shape> AsShape(ExtentExpr &&arrayExpr);
```

Comment:
What does it mean when this returns a nullopt as opposed to an empty Shape?

Reply:
A scalar has a valid shape that happens to be empty.  A failure to discern a shape at compilation time results in an absent value.